### PR TITLE
WIP: feat/pubsub

### PR DESCRIPTION
**Pull request by diasdavid**
> **From @haadcode https://github.com/ipfs/js-ipfs-api/pull/470**

This PR will add complete go-ipfs Pubsub API.

Missing:
- Documentation

Dependencies:
- https://github.com/ipfs/js-ipfs-api/pull/465
- Need to have go-ipfs 0.4.5-x released so that there's a binary that can be used with this api

### CODE REVIEW

**Comment by diasdavid** [comment]
@haadcode having troubles with this PR, now that I added (manually) the missing test files, it is missing other modules. Could you fetch this branch and try to run yourself and see which files are missing? I'm getting it was just the original PR that was missing things and you have the code in your machine.

**Comment by diasdavid** [comment]
@haadcode also, can you make a PR to js-ipfsd-ctl with the changes you made?

File: src/api/pubsub.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+      }
+
+      if (!options) {
+        options = defaultOptions
+      }
+
+      // If we're already subscribed, return an error
+      if (subscriptions[topic]) {
+        return callback(new Error(`Already subscribed to '${topic}'`))
+      }
```
Comment (diasdavid) [inline src/api/pubsub.js]: If we have several subscriptions, we just need to feed the messages to all of them.

File: src/api/pubsub.js [hunk 2]
```diff
@@ -57,11 +57,13 @@ module.exports = (send) => {
       }
     },
     publish: promisify((topic, data, callback) => {
-      const buf = Buffer.isBuffer(data) ? data : new Buffer(data)
+      if (!Buffer.isBuffer(data)) {
+        return callback(new Error('data must be a Buffer'))
```
Comment (diasdavid) [inline src/api/pubsub.js]: <3

File: package.json [hunk 3]
```diff
@@ -64,7 +64,7 @@
     "gulp": "^3.9.1",
     "hapi": "^16.0.1",
     "interface-ipfs-core": "^0.22.1",
-    "ipfsd-ctl": "^0.17.0",
+    "@haad/ipfsd-ctl": "^0.18.0-beta.5",
```
Comment (diasdavid) [inline package.json]: Needs to be updated

**Comment by dignifiedquire** [comment]
Closing in favor of #493

**Comment by haadcode** [comment]
Remind me @dignifiedquire @diasdavid that we need to talk about strategy/process for releasing dev versions in the future and handling dependencies in a sane way.

**Pull request closed without merging**